src/webview/web_context.rs

```diff
@@ -170,29 +177,174 @@ pub mod unix {
     fn context(&self) -> &WebContext;
 
+    /// The GTK [`UserContentManager`] of all webviews in the context.
+    fn manager(&self) -> &UserContentManager;
+
+    /// Register a custom protocol to the web context
+    fn register_uri_scheme<F>(
+      &self,
+      name: &str,
+      handler: F,
+      window: Rc<Window>,
+    ) -> crate::Result<()>
+    where
+      F: Fn(&Window, &str) -> crate::Result<(Vec<u8>, String)> + 'static;
+
+    /// Add a [`WebView`] to the queue waiting to be opened.
+    ///
+    /// Using the queue prevents data race issues with loading uris for multiple [`WebView`]s in the
+    /// same context at the same time. Occasionally, the one of the [`WebView`]s will be clobbered
+    /// and it's content will be injected into a different [`WebView`].
+    ///
+    /// Example of `webview-c` clobbering `webview-b` while `webview-a` is okay:
+    /// ```text
+    /// webview-a triggers load-change::started
+    /// URISchemeRequestCallback triggered with webview-a
+    /// webview-a triggers load-change::committed
+    /// webview-a triggers load-change::finished
+    /// webview-b triggers load-change::started
+    /// webview-c triggers load-change::started
+    /// URISchemeRequestCallback triggered with webview-c
+    /// URISchemeRequestCallback triggered with webview-c
+    /// webview-c triggers load-change::committed
+    /// webview-c triggers load-change::finished
+    /// ```
+    ///
+    /// In that example, `webview-a` will load fine. `webview-b` will remain empty as the uri was
+    /// never loaded. `webview-c` will contain the content of both `webview-b` and `webview-c`
+    /// because it was triggered twice even through only started once. The content injected will not
+    /// be sequential, and often is interjected in the middle of one of the other contents.
+    fn queue_load_uri(&self, webview: Rc<WebView>, url: Url);
+
+    /// Flush all queued [`WebView`]s waiting to load a uri.
+    fn flush_queue_loader(&self);
+
     /// If the context allows automation.
     ///
     /// **Note:** `libwebkit2gtk` only allows 1 automation context at a time.
     fn allows_automation(&self) -> bool;
   }
 
   impl WebContextExt for super::WebContext {
-    fn app_info(&self) -> &ApplicationInfo {
-      &self.os.app_info
-    }
-
     fn context(&self) -> &WebContext {
       &self.os.context
     }
 
+    fn manager(&self) -> &UserContentManager {
+      &self.os.manager
+    }
+
+    fn register_uri_scheme<F>(
+      &self,
+      name: &str,
+      handler: F,
+      window: Rc<Window>,
+    ) -> crate::Result<()>
+    where
+      F: Fn(&Window, &str) -> crate::Result<(Vec<u8>, String)> + 'static,
+    {
+      let mut protocols = self
+        .os
+        .registered_protocols
+        .lock()
+        .expect("poisoned protocol hashset");
+      let exists = !protocols.insert(name.to_string());
+      drop(protocols);
+
+      if exists {
+        Err(Error::DuplicateCustomProtocol(name.to_string()))
```

Comment:
The conflict of duplocate custom protocol only happen if they share same web context.
I think this should let users to decide and we could add some warnings on the behaviour of this.

Reply:
yeah, in my item #2 the `with_custom_protocol` returns a `Result` like some of the other builder items so that the user can choose to ignore it instead of it popping off in the InnerWebView constructor. I think we can support custom protocols on both the `WebviewBuilder` and the `WebContext`, where using `WebviewBuilder` it will just swallow this result inside the webview constructor, and from `WebContext` it allows the user to handle it how they want

Reply:
I went with `register_uri_scheme` and `try_register_uri_scheme`. Both will return an `Err(DuplicateCustomProtocol)` when a duplicate is passed, but the non-try one will still pass it to the underlying platform (only webkit2gtk) in this instance. The code in `InnerWebview` has been updated to keep the old behavior. I'll leave this conversation open if you think we should resolve it another way.

Reply:
I think it's fine to add both methods. But for adding URI method to `WebContext` should be exclusive on Linux imho.
Mac and Windows handle them differently (and it's even a hack on Window for now). They probably don't need it.